### PR TITLE
SpikeInterface integration (WIP)

### PR DESCRIPTION
**Pull request by JaerongA**

### CODE REVIEW

File: element_array_ephys/ephys_no_curation.py [hunk 1]
```diff
@@ -1168,177 +1168,245 @@ class Waveform(dj.Part):
...
+            waveforms_folder = [
+                f for f in output_dir.parent.rglob(r"*/waveforms*") if f.is_dir()
+            ]
+
+            if (output_dir / "mean_waveforms.npy").exists():
+                unit_waveforms = np.load(
+                    output_dir / "mean_waveforms.npy"
+                )  # unit x channel x sample
+
+                def yield_unit_waveforms():
```
Comment (JaerongA) [inline element_array_ephys/ephys_no_curation.py:1264]: @ttngu207 I didn't touch this `else` block, but why is this function duplicated?

**Review by kushalbakshi — COMMENTED** [review]
Thanks @JaerongA. Most of this looks good. A few minor suggestions for docstrings before we do the full review this afternoon. It's probably a good idea to run black formatting on all of these files before then too.

**Comment by JaerongA** [comment]
> Thanks @JaerongA. Most of this looks good. A few minor suggestions for docstrings before we do the full review this afternoon. It's probably a good idea to run black formatting on all of these files before then too.

Sounds good. It should have already been black-formatted but let me double check.

File: element_array_ephys/ephys_no_curation.py [hunk 2]
```diff
@@ -956,81 +962,188 @@ class Unit(dj.Part):
...
                     }
                 )
 
+        else:
+            kilosort_dataset = kilosort.Kilosort(output_dir)
+            acq_software, sample_rate = (EphysRecording & key).fetch1(
+                "acq_software", "sampling_rate"
+            )
+
+            sample_rate = kilosort_dataset.data["params"].get(
```
Comment (JaerongA) [inline element_array_ephys/ephys_no_curation.py:1078]: why is this being fetched twice?

File: element_array_ephys/ephys_no_curation.py [hunk 3]
```diff
@@ -327,23 +326,26 @@ def make(self, key):
                 break
         else:
             raise FileNotFoundError(
-                "Ephys recording data not found!"
+                f"Ephys recording data not found! for {key}."
```
Comment (ttngu207) [inline element_array_ephys/ephys_no_curation.py:329]:
```suggestion
                f"Ephys recording data not found in {session_dir}."
```

File: element_array_ephys/ephys_no_curation.py [hunk 4]
```diff
@@ -352,36 +354,101 @@ def make(self, key):
...
+                    probe.ElectrodeConfig.Electrode.insert(
+                        {**electrode_config_key, **electrode}
+                        for electrode in electrode_group_members
+                    )
 
             self.insert1(
                 {
                     **key,
-                    **generate_electrode_config(probe_type, electrode_group_members),
+                    "electrode_config_hash": electrode_config_hash,
```
Comment (ttngu207) [inline element_array_ephys/ephys_no_curation.py:402]: Why removing this `generate_electrode_config` helper function only to have to duplicate this code block multiple time?

File: element_array_ephys/ephys_no_curation.py [hunk 5]
```diff
@@ -1644,99 +1765,6 @@ def get_openephys_probe_data(ephys_recording_key: dict) -> list:
...
-
-        channel2electrode_map = {
-            channel_idx: probe_electrodes[channel_idx]
-            for channel_idx in probe_dataset.ap_meta["channels_indices"]
-        }
-
-    return channel2electrode_map
-
-
-def generate_electrode_config(probe_type: str, electrode_keys: list) -> dict:
```
Comment (ttngu207) [inline element_array_ephys/ephys_no_curation.py:1432]: I think it makes sense to keep this function, as it is reusable in may places

File: element_array_ephys/ephys_no_curation.py [hunk 6]
```diff
@@ -1644,99 +1765,6 @@ def get_openephys_probe_data(ephys_recording_key: dict) -> list:
     return probe_data
 
 
-def get_neuropixels_channel2electrode_map(
```
Comment (ttngu207) [inline element_array_ephys/ephys_no_curation.py:1383]: Makes sense to get rid of this

**Review by ttngu207 — CHANGES_REQUESTED** [review]

File: element_array_ephys/ephys_no_curation.py (hunk 5 above)
Comment (JaerongA) [reply element_array_ephys/ephys_no_curation.py:1432]: @ttngu207. I've made this [commit](https://github.com/datajoint/element-array-ephys/pull/179/commits/1d805849f20ee1e5ce6a911d06d189f9d699900f) to address your comment. It's true that there's some code duplication but I think this enhances separation of concern (compute vs. insert) and readability. Previously it wasn't very clear what that function does and where probe.ElectrodeConfig was being populated. I'm open to any suggestions though.

Comment (ttngu207) [reply element_array_ephys/ephys_no_curation.py:1432]: Right, it wasn't obvious where/how entries for `probe.ElectrodeConfig` were generated/inserted.
I agree with this new change. I'd suggest to go a bit further and update the `generate_electrode_config_name` function to be `generate_electrode_config_entry` - returning `(econfig_entry, econfig_electrodes)` (including the hash)

Then you just use that for the insert, making the insert more visible without duplicating a lot of code

```
if not probe.ElectrodeConfig & {"electrode_config_hash": econfig_entry['electrode_config_hash']}:
        probe.ElectrodeConfig.insert1(econfig_entry)
        probe.ElectrodeConfig.Electrode.insert(econfig_electrodes)
 ```
 (something like that)

Comment (JaerongA) [reply element_array_ephys/ephys_no_curation.py:1432]: @ttngu207 Sounds good. I've updated the function.

**Review by ttngu207 — APPROVED** [review]

**Pull request merged**